### PR TITLE
NON-287: Disallow adding a non-associations to prisoners with null locations

### PR DESCRIPTION
**Pull request by ushkarev**
The api now responds with 409 CONFLICT when trying to create a non-association with a prisoner with a null location.

Also, allow for offender search results that have no prison name or ID. Because those fields _can_ be null, we are forced to propagate the nullability into _this_ api.

**It does mean that our api contract changes!**

Related: [ui#184](https://github.com/ministryofjustice/hmpps-non-associations/pull/184) & [ui#185](https://github.com/ministryofjustice/hmpps-non-associations/pull/185)

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/offendersearch/OffenderSearchPrisoner.kt
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/offendersearch/OffenderSearchPrisoner.kt:10]: this is the main change: all location-type fields are nullable. the rest of the changes in this commit propagate these nulls

**Review by Mjwillis — APPROVED** [review]
LGTM

File: clients/node/README.md
Comment (xoen) [inline clients/node/README.md:113]: Why tho? A user knows what's doing no? They searched for this individual and they decided to add this person to this non-association.

It seems to me we're trying to be too clever with the logic here. The user probably have more context than the API.

The person may also get a booking soon and the user reasons to add a non-association to this person may be sound?

File: clients/node/src/errorTypes.ts
Comment (xoen) [inline clients/node/src/errorTypes.ts:5]: I kinda liked the direct link to the code.
Although sure code change over time, pin the commit hash is an idea but then it may be stale code...so maybe that's why you made the link more generic?

Comment (ushkarev) [reply clients/node/src/errorTypes.ts:5]: problem is i'm changing `ErrorCode` here so cannot link to the un-merged code… so i thought i might give up on these direct links. what do you think is better 1) drop all direct commit links 2) link to `main` and cross fingers that if something moves, we remember to repoint these links? 3) wait until merge to put in direct commit links

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/offendersearch/OffenderSearchPrisoner.kt
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/offendersearch/OffenderSearchPrisoner.kt:10]: Wouldn't this have an impact on (other) clients tho? It sounds like a breaking change (although a rare one as I'd imagine most of the data _will_ have the prisoner number)

File: clients/node/src/errorTypes.ts
Comment (xoen) [reply clients/node/src/errorTypes.ts:5]: I think this is probably the best compromise: Comment has class path/packages and link is to repo.

**Review by xoen — APPROVED** [review]
I'm not convinced we should try to be clever on the API side and prevent creating a non-association just because one of the people involved doesn't have a prisonId.

It should be a rare occasions anyway. But changes look fine.

PS: I think the nullable may be potentially break some of the clients (sometimes). But maybe it's so rare we're ok with this tradeoff?

**Pull request merged**
